Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,413 @@
+        ///  |---------|
+        /// </summary>
+        private float CalculateFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                Quaternion inverseRotation = Quaternion.Inverse(m_BoundingShape2D.transform.rotation);
+                Vector3 planePosition = inverseRotation * m_BoundingShape2D.transform.position;
+                Vector3 cameraPosition = inverseRotation * vcam.transform.position;
+                float distance = Mathf.Abs(planePosition.z - cameraPosition.z);
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+            return frustumHeight;
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine)
+        {
+            if (ShrinkablePolygon.IsInside(m_currentPathCache, positionToConfine))
+            {
+                return Vector2.zero;
+            }
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < m_currentPathCache.Count; ++i)
+            {
+                int numPoints = m_currentPathCache[i].Count;
+                if (numPoints > 0)
+                {
+                    Vector2 v0 = m_currentPathCache[i][numPoints - 1];
+                    for (int j = 0; j < numPoints; ++j)
+                    {
+                        Vector2 v = m_currentPathCache[i][j];
+                        Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                        float distance = Vector2.SqrMagnitude(positionToConfine - c);
+                        if (distance < minDistance)
+                        {
+                            minDistance = distance;
+                            closest = c;
+                        }
+                        v0 = v;
+                    }
+                }
+            }
+            return closest - positionToConfine;
+        }
+        
+        private class VcamExtraState
+        {
+            public Vector3 m_previousDisplacement;
+            public float confinerDisplacement;
+            public bool applyAfterAim;
+        };
+
+        private float m_sensorRatioCache;
+        private Vector3 m_boundingShapePositionCache;
+        private Vector3 m_boundingShapeScaleCache;
+        private Quaternion m_boundingShapeRotationCache;
+        /// <summary>
+        /// Invalidates path cache.
+        /// </summary>
+        private void InvalidatePathCache()
```

Comment:
I'm confused by all the different levels of cache.  I'd like to see all the caches collected into 2 classes (or structs), each with a clear API for validation and invalidation.

ShapeCache: contains all state that's dependent only on the settings in the confiner: bounding shape, shrinkToPoint, maxOrthoSize.  Contains nothing that is dependent on anything in the vcam itself (except maybe aspect ratio, which we can assume to be constant among vcam children).

VcamShapeCache (lives inside VcamExtraState): contains all the cache items that are dependent on something in the vcam (e.g. orthoSize).